CLOUDP-132397 Add evergreen task to upload the release tarball to an s3 bucket

visualzhou opened the pull request:
This `compile` task archives the binary and the hot-restart python script to a tarball and uploads the tarball to a directory under `mongodb-mms-build-envoy-serverless/build/${build_id}`. The `push` task copies the tarball to `mongodb-mms-build-envoy-serverless/tarballs/` which will be consumed by the promotion script tracked by CLOUDP-132402. Separating compile and push allows patch builds to skip `push`, otherwise tarball file names from patch builds on the same base commit will inevitably conflict under the `tarballs` forlder, which isn't a big deal but misleading.

The S3 bucket `mongodb-mms-build-envoy-serverless` allows public download for a given object, so that the file links on the `compile` evergreen task could work.

A new user [`mms-build-envoy-serverless`](https://us-east-1.console.aws.amazon.com/iam/home#/users/mms-build-envoy-serverless) is created under the `MMS Engineering Build` AWS account and the corresponding policy for the above S3 bucket has been set. Its credentials have been added to the `envoy-serverless` Evergreen project settings under `release_aws_key` and `release_aws_secret`.

milkie commented on evergreen/archive-evergreen.sh:
```diff
@@ -0,0 +1,27 @@
+VERSION=$(git describe --tags --dirty --always --match 'v[0-9]*' | cut -c 2-)
+cd -
+
+# The archives has to be put in a directory because evergreen s3.put will scan the working directory
```
```suggestion
# The archives have to be put in a directory because evergreen s3.put will scan the working directory
```

milkie commented on evergreen/archive-evergreen.sh:
```diff
@@ -0,0 +1,27 @@
+# working directory to "ARCHIVE_PATH" is a workaround.
+ARCHIVE_PATH=archive
+# Example archive file name: envoy-serverless-1.21.3-4-gdbdcfa34cf.rhel7.amd64.tar.gz
+ARCHIVE_DIR=$ARCHIVE_PATH/envoy-serverless-${VERSION}.rhel7.amd64
```
Does it make sense that rhel7 and amd64 are hardcoded here?  Should we name the script file with "rhel7" in the name maybe?

visualzhou replied:
Keeping this file generic and using Evergreen expansions would be ideal, but since we only have one build variant here, I hardcoded it. There will be other work to support ARM in the future, do you mind leaving the change to that time?

milkie replied:
I think we don't need to genericize this script, but it is still prudent to name it with "rhel7" so we remember it's platform-specific.  evergreen/archive-evergreen-rhel7-amd64.sh ?

visualzhou replied:
Got it. Done!

lydiastepanek commented on evergreen/evergreen.yml:
```diff
@@ -25,18 +33,66 @@ functions:
-    release: 7
-    image: centos:7
-    arch: x86_64
   tasks:
```
Should we run `compile` and `test` tasks on PR merge, and the `push` task on merge to master? Looks like we could use something like `patchable: false` to configure `push` to only run on merge: https://github.com/evergreen-ci/evergreen/wiki/Project-Configuration-Files#limiting-when-a-task-will-run

visualzhou replied:
Running patches on PRs needs more config on github. Since we won't develop in this repo a lot, I didn't set it.

Great point on `patchable: false` though. I needed patch builds to test my changes, but we should set it. Done.

lydiastepanek replied:
Seems like you can just click this button in the evergreen project settings ([source](https://github.com/evergreen-ci/evergreen/wiki/Using-Repo-Level-Settings#how-it-works))
![Screen Shot 2023-01-04 at 9 17 43 AM](https://user-images.githubusercontent.com/7087489/210574901-739fb31a-7a73-4132-9319-0d150c844949.png)
> Running patches on PRs needs more config on github.

visualzhou replied:
Thanks! That's great to know. Will do.

visualzhou commented:
Addressed comments. Here's the patch build without the last commit about "patchable".
https://spruce.mongodb.com/version/63b4d5e50ae6064969a665b2/tasks?sorts=STATUS%3AASC%3BBASE_STATUS%3ADESC

JamesHeppenstall reviewed (commented):
Code changes LGTM but had a couple questions about the Evergreen tasks:
1. For the test task, I noticed that some tests are flaky and that we pass in `--flaky_test_attempts=//test/integration.*@3` to `bazel test ...`. I see that this was copied over from the RPM spec. Do we know why these tests are flaky?
2. How do I check that the `push` task has successfully copied over the tarball?

visualzhou commented:
> Code changes LGTM but had a couple questions about the Evergreen tasks:
> 
> 1. For the `test` task, I noticed that some tests are flaky and that we pass in `--flaky_test_attempts=//test/integration.*@3` to `bazel test ...`. I see that this was copied over from the RPM spec. Do we know why these tests are flaky?
> 2. How do I check that the `push` task has successfully copied over the tarball?

1. I don't know why they are flaky. We could revisit it after moving to Amazon Linux 2 (or 2022?). I don't think CentOS 7 is tested in Envoy CI.
3. The log line of "push" shows it's copied. Besides, I checked the target S3 bucket manually.
```
copy: s3://mongodb-mms-build-envoy-serverless/build/envoy_serverless_centos7_patch_dbdcfa34cf779309944693b26bd1589a98f035ce_63b4d5e50ae6064969a665b2_23_01_04_01_27_15/envoy-serverless-1.21.3-4-gdbdcfa34cf-dirty.rhel7.amd64.tar.gz to s3://mongodb-mms-build-envoy-serverless/tarballs/envoy-serverless-1.21.3-4-gdbdcfa34cf-dirty.rhel7.amd64.tar.gz
```

visualzhou commented:
@lydiastepanek it turns out Evergreen requires setting an evergreen alias for the variants and tests to run on PRs in order to enable automatic testing. Usually that requires defining task tags. We will set it up for the `serverless-proxy` repo, but I don't think it's worth the effort to automate the workflow on this project more right now.

lydiastepanek commented:
@visualzhou Ok! Makes sense
